WordPress/Classes/ViewRelated/Domains/Domain registration/DomainPurchasingWebFlowController.swift

```diff
@@ -0,0 +1,338 @@
+import UIKit
+import AutomatticTracks
+import Sentry
+
+final class DomainPurchasingWebFlowController {
+
+    // MARK: - Constants
+
+    fileprivate enum Constants {
+        static let checkoutWebAddress = "https://wordpress.com/checkout/"
+        static let storeSandboxCookieName = "store_sandbox"
+        static let storeSandboxCookieDomain = ".wordpress.com"
+        static let checkoutSuccessURLPrefix = "https://wordpress.com/checkout/thank-you/"
+        static let checkoutURLPrefix = "https://wordpress.com/checkout"
+    }
+
+    // MARK: - Dependencies
+
+    /// The view controller that presents the domain checkout web page.
+    weak private var presentingViewController: UIViewController?
+
+    /// The service that interacts with the Backend API.
+    private let shoppingCartService: RegisterDomainDetailsServiceProxyProtocol
+
+    /// Provides an API to capture errors.
+    private let crashLogger: CrashLogging
+
+    // MARK: - Execution Variables
+
+    /// Set when a domain checkout web page is presented.
+    private weak var presentedViewController: UINavigationController?
+
+    /// Observe url changes and it is set when a domain checkout web page is presented.
+    private var webViewURLChangeObservation: NSKeyValueObservation?
+
+    // MARK: - Init
+
+    init(viewController: UIViewController,
+         shoppingCartService: RegisterDomainDetailsServiceProxyProtocol = RegisterDomainDetailsServiceProxy(),
+         crashLogger: CrashLogging = .main) {
+        self.presentingViewController = viewController
+        self.shoppingCartService = shoppingCartService
+        self.crashLogger = crashLogger
+    }
+
+    // MARK: - API
+
+    func purchase(domain: FullyQuotedDomainSuggestion, site: Blog, completion: CompletionHandler? = nil) {
+        purchase(domain: domain.remoteSuggestion(), site: site, completion: completion)
+    }
+
+    func purchase(domain: DomainSuggestion, site: Blog, completion: CompletionHandler? = nil) {
+        let middleware: CompletionHandler = { [weak self] result in
+            if let self = self, case let .failure(error) = result, !error.trusted {
+                let userInfo = self.userInfoForError(error, domain: domain, site: site)
+                self.crashLogger.logError(error, userInfo: userInfo, level: error.level)
+            }
+            if let completion {
+                DispatchQueue.main.async {
+                    completion(result)
+                }
+            }
+        }
+        guard let presentingViewController = self.presentingViewController else {
```

Comment:
We can probably write it as 
```Swift
guard let presentingViewController else {
```

It could also make more sense to guard this at the beginning of the function

Reply:
Resolved in 39e236e508729f7d37d3077520a9ba9f7750f6e2